### ISSUE TITLE
Mail Composer: Newline vs. paragraph

### ISSUE THREAD

**Issue by benbucksch** [mustang-im/mustang#35]
* In HTML, a paragraph has vertical whitespace (= an empty line = bottom margin) between paragraphs.
* In editors, when users press Enter key, they expect a simple newline, not 2 newlines.
* In TipTap, when you press Enter key, it creates a new paragraph, including the unexpected empty newline.
* I am currently working around that by styling a `<p>` to not have a bottom margin. This works only when the recipient also renders CSS.

### Reproduction
* In mail composer
* Type 1
* Hit Enter
* Type 2
* Hit Enter
* Hit Enter
* Type 3
* Send

### Actual result
(without the CSS workaround)

1. Every Enter creates a `<p>`.

HTML:
```html
<p>1</p>
<p>2</p>
<p></p>
<p>3</p>
```

Display:
```
1

2


3
```

### Expected result

1. The first Enter creates a `<br>`.
2. Two Enter successively (= in a row) create a `<p>`.
3. Third Enter creates `<br>` within the `<p>`,
4. the forth Enter creates a second `<p>`, and so on.

HTML:
```html
<p>1<br>
2</p>
<p>3</p>
```

Display:
```
1
2

3
```

### Process
* You need to first remove my CSS workaround, in file `content.css`.
* If `<br>` is not yet allowed, you need to whitelist this HTML element.
* Create a TipTap extension which implements the above behavior.
* It has to harmonize with #32 (in both modes), #30 and other extensions we have.
* Test it with existing documents (e.g. pasting them from websites)

**Comment by benbucksch** [comment]
@jermy-c Can you please pick this up? This is a release blocker now.

**Comment by benbucksch** [comment]
Basically, if the cursor is right before or after a `<br/>`, and the user presses ENTER, then remove the `<br/>` and replace it with `</p><p>`.

**Comment by benbucksch** [comment]
FWIW (just in case it's relevant here), Links `<a href>` should stop once you press enter.

**Comment by benbucksch** [comment]
Check what Thunderbird does and sends out as HTML.
Make sure that <p> is rendered with newline in other email clients.

**Comment by benbucksch** [comment]
### Reproduction
* In the mail composer, in HTML, enter text, and send it to yourself
```
1 line
2 line
3 line

4 line new p
5 line
6 line

7 after new p


8 after 2 new p
```

### Actual result
1. Plaintext: Linebreaks are doubled
```
1 line

2 line

3 line



4 line new p

5 line

6 line



7 after new p





8 after 2 new p
```

2. HTML
```html
<html><head></head><body>
<p>1 line</p>
<p>2 line</p>
<p>3 line</p>
<p></p>
<p>4=
 line new p</p>
<p>5 line</p>
<p>6 line</p>
<p></p>
<p>7 after new =
p</p>
<p></p>
<p></p>
<p>8 after 2 new p</p>
<p></p>
<p></p>
```

### Expected result
1. Plaintext: Looks exactly like what you typed:
```
1 line
2 line
3 line

4 line new p
5 line
6 line

7 after new p


8 after 2 new p
```

2. HTML: Looks exactly like what you typed:
```html
<html><head></head><body>
<p>1 line<br/>
<p>2 line<br/>
<p>3 line</p>

<p>4 line new p<br/>
<p>5 line<br/>
<p>6 line<br/>

<p>7 after new p</p>
<p><br/></p>
<p>8 after 2 new p</p>
```

**Comment by benbucksch** [comment]
Within lists (`<ol><li>`, `<ul><li>`), use the default behavior. When you press SHIFT+ENTER, it should insert a `<br/>`. `<p>` within `<li>` is not needed - whatever is easier.

**Comment by benbucksch** [comment]
With the PR #510

### Test case
I've tested it and it works really well. What works:
* First ENTER creates `<br>` :+1: 
* Second consequtive ENTER removes `<br>` and creates `<p>` :+1:
* Bulletet list works :+1:
* ENTER in bulleted list creates new list item  :+1:  
* SHIFT-ENTER in bulleted list creates `<br>`  :+1: 
* Second SHIFT-ENTER in bulleted list creates second `<br>` :+1: 
* ENTER twice in bulleted list ends the bulleted list and creates a normal `<p>`  :+1: 

### Visual results
The result
* looks correct in our editor (after my stylesheet fixes, please check)
* looks correct in our mail viewer
* looks correct in our plaintext mail viewer (top right [T] button)
* looks correct in Thunderbird's mail viewer
* looks correct in Thunderbird's plaintext mail viewer (menu | View | Message body as | plain text)
* looks correct when replying with our editor
* looks correct when replying with Thunderbird's HTML editor
* looks correct when replying with Thunderbird's plaintext editor (SHIFT-click on Reply button in Thunderbird), apart from bulleted list items that have excessive newlines, see below.

### Mail source results
The mail source code: (message menu [...] | View Source)
* is correct in HTML. The HTML elements are exactly as specced above. There is no stylesheet needed. :+1: 
* is correct in plaintext.

**Comment by benbucksch** [comment]
### Problems
The only problems I could find were with bulleted lists:
1. List items contain `<p>`, and visible margins between list items, but they should not. I get `<ul><li><p>Text</p></li></ul>`. Expected: `<ul><li>Text</li></ul>`.
2. Consequence: That is also the reason why you get excessive newlines when replying in the plaintext editor in Thunderbird. Reason: Thunderbird must be down-converting the HTML before going into the plaintext editor, instead of replyintg to the plaintext straight.
3. Text ENTER `* List item` should create a bulleted list, but it does not. Text ENTER ENTER `* List item` works as expected. That means bulleted lists are not created after a `<br>`, only after a `<p>`.

Can you please fix those 2 problems in a new PR?

**Comment by benbucksch** [comment]
Parula HTML 👍 
![Image](https://github.com/user-attachments/assets/eea31010-7d6a-4ad5-aea2-418461a1816c)

Parula Plaintext 👍 
![Image](https://github.com/user-attachments/assets/73e9d620-e2ac-4ed9-9aa2-cee564ed8674)

Thunderbird HTML 👍 
![Image](https://github.com/user-attachments/assets/f2b1ac48-db9b-4a5f-9614-cb8a1af5c0a9)

Thunderbird plaintext 👍 
![Image](https://github.com/user-attachments/assets/dd7d0e55-c5b2-4330-9a45-81a074713010)

Thunderbird Reply HTML 👍 
![Image](https://github.com/user-attachments/assets/2db8a79c-1e78-4bcb-a429-4a953458e94e)

Thunderbird Reply plaintext ❌ 
![Image](https://github.com/user-attachments/assets/f7f5cdfc-4a1d-4a69-b307-7579e4c0b935)

Parula Reply HTML 👍 
![Image](https://github.com/user-attachments/assets/ec1b7ed7-5361-4576-a3f7-55f37fb7ae24)

**Comment by jermy-c** [comment]
> 1. List items contain `<p>`, and visible margins between list items, but they should not. I get `<ul><li><p>Text</p></li></ul>`. ❌ Expected: `<ul><li>Text</li></ul>`.

I don't think is could easily be fixed: https://github.com/ueberdosis/tiptap/issues/118. Because everything in Tiptap/Prosemirror needs to be wrapped in a `<p>`

> Text ENTER `* List item` should create a bulleted list, but it does not. ❌ Text ENTER ENTER `* List item` works as expected. That means bulleted lists are not created after a `<br>`, only after a `<p>`.

Ok, this one might be simpler to fix since the RegularExpression checks for the the start of the text node instead we should check the start of the new line.

**Comment by jermy-c** [comment]
Attempts:
- Change the inputRule regex to check the start of a newline instead of start of a node/textblock. This doesn't work because it just wraps the entire node.

```html
<p>
text<br>
- list item
</p>
```

Becomes

```html
<ul>
<li><p>textlistitem</p></li>
</ul>
```

- Creating a custom inputRule handler. There's no constructor exported for creating custom Input Handlers and also you need some of the `@tiptap/core` functions that aren't exported. https://github.com/ueberdosis/tiptap/blob/develop/packages/core/src/inputRules/wrappingInputRule.ts

- For toggling list wraps. I haven't attempted it but it could be possible if you can write a customCommand to split the node where `<br>` is present before wrapping but it'll also need to be able to wrap multiple nodes in that case. I'll need to look into it.

**Comment by jermy-c** [comment]
Attempted creating the inputRule handler and custom Toggling command but still no success. I'm not sure how to get the node selection of the `<p>` and do a querySelect on the node.

Code on branch `jeremy/list-conventions`: https://github.com/mustang-im/mustang/tree/jeremy/list-conventions

**Comment by benbucksch** [comment]
Note: It's not just the toolbar button. You can also create bulleted lists by typing markdown:
ENTER * SPACE Text
```
Text
* dsfrgdhdh
* sdfsf
```

Same for numbered lists

**Comment by benbucksch** [comment]
I've filed #532 for the bulleted lists discussed in the last few comments.
